File: frontend/flutter_application_1/lib/services/recipe_discovery_service.dart
import 'package:dio/dio.dart';
import 'package:flutter/foundation.dart';

import '../models/recipe_models.dart';
import '../utils/app_constants.dart';

class RecipeDiscoveryService {
  late final Dio _dio;
  static const bool _useMockData = false; // Use real backend API

  RecipeDiscoveryService() {
    _dio = Dio(BaseOptions(
      baseUrl: AppConstants.baseUrl,
      connectTimeout: const Duration(seconds: 30),
      receiveTimeout: const Duration(seconds: 30),
      headers: {
        'Content-Type': 'application/json',
        'Accept': 'application/json',
      },
    ));

    // Add interceptors for logging and error handling
    if (kDebugMode) {
      _dio.interceptors.add(LogInterceptor(
        requestBody: true,
        responseBody: true,
        logPrint: (obj) => debugPrint(obj.toString()),
      ));
    }

    // Add auth interceptor
    _dio.interceptors.add(InterceptorsWrapper(
      onRequest: (options, handler) {
        // Auth token will be set by the provider
        handler.next(options);
      },
      onError: (error, handler) {
        debugPrint('Recipe Discovery API Error: ${error.message}');
        handler.next(error);
      },
    ));
  }

  // Set authorization token
  void setAuthToken(String token) {
    _dio.options.headers['Authorization'] = 'Bearer $token';
  }

  // Clear authorization token
  void clearAuthToken() {
    _dio.options.headers.remove('Authorization');
  }

  /// Search recipes with comprehensive filtering and pagination
  Future<RecipeSearchResult> searchRecipes({
    String searchQuery = '',
    RecipeFilters? filters,
    int page = 1,
    int limit = 20,
    String sortBy = 'name',
    String sortOrder = 'asc',
  }) async {
    if (_useMockData) {
      return _getMockSearchResult(searchQuery, filters, page, limit, sortBy, sortOrder);
    }

    try {
      Map<String, dynamic> queryParams = {
        'page': page.toString(),
        'limit': limit.toString(),
        'sort_by': sortBy,
        'sort_order': sortOrder,
      };

      if (searchQuery.isNotEmpty) {
        queryParams['q'] = searchQuery;
      }

      if (filters != null) {
        queryParams.addAll(filters.toQueryParams());
      }

      final response = await _dio.get(
        '/api/v1/recipes/search',
        queryParameters: queryParams,
      );

      if (response.statusCode == 200) {
        return RecipeSearchResult.fromJson(response.data);
      } else {
        throw Exception('Failed to search recipes: ${response.statusCode}');
      }
    } on DioException catch (e) {
      debugPrint('Recipe search error: ${e.message}');
      // Fall back to mock data on error
      return _getMockSearchResult(searchQuery, filters, page, limit, sortBy, sortOrder);
    } catch (e) {
      debugPrint('Unexpected error in recipe search: $e');
      // Fall back to mock data on error
      return _getMockSearchResult(searchQuery, filters, page, limit, sortBy, sortOrder);
    }
  }

  /// Get detailed information for a specific recipe
  Future<Recipe> getRecipeDetails(String recipeId) async {
    if (_useMockData) {
      return _getMockRecipe(recipeId);
    }

    try {
      final response = await _dio.get('/api/v1/recipes/$recipeId');

      if (response.statusCode == 200) {
        return Recipe.fromJson(response.data);
      } else {
        throw Exception('Failed to get recipe details: ${response.statusCode}');
      }
    } on DioException catch (e) {
      debugPrint('Recipe details error: ${e.message}');
      // Fall back to mock data on error
      return _getMockRecipe(recipeId);
    } catch (e) {
      debugPrint('Unexpected error getting recipe details: $e');
      // Fall back to mock data on error
      return _getMockRecipe(recipeId);
    }
  }

  /// Get available filter options for the recipe catalog
  Future<FilterOptions> getFilterOptions() async {
    if (_useMockData) {
      return _getMockFilterOptions();
    }

    try {
      final response = await _dio.get('/api/v1/recipes/filters/options');

      if (response.statusCode == 200) {
        return FilterOptions.fromJson(response.data);
      } else {
        throw Exception('Failed to get filter options: ${response.statusCode}');
      }
    } on DioException catch (e) {
      debugPrint('Filter options error: ${e.message}');
      return _getMockFilterOptions();
    } catch (e) {
      debugPrint('Unexpected error getting filter options: $e');
      return _getMockFilterOptions();
    }
  }

  /// Get search suggestions based on partial query
  Future<List<String>> getSearchSuggestions(String query, {int limit = 10}) async {
    if (_useMockData) {
      return _getMockSearchSuggestions(query, limit);
    }

    try {
      if (query.length < 2) {
        return [];
      }

      final response = await _dio.get(
        '/api/v1/recipes/suggestions',
        queryParameters: {
          'q': query,
          'limit': limit.toString(),
        },
      );

      if (response.statusCode == 200) {
        final data = response.data as Map<String, dynamic>;
        return (data['suggestions'] as List<dynamic>).cast<String>();
      } else {
        throw Exception('Failed to get search suggestions: ${response.statusCode}');
      }
    } on DioException catch (e) {
      debugPrint('Search suggestions error: ${e.message}');
      return _getMockSearchSuggestions(query, limit);
    } catch (e) {
      debugPrint('Unexpected error getting search suggestions: $e');
      return _getMockSearchSuggestions(query, limit);
    }
  }

  /// Get trending/popular recipes
  Future<List<Recipe>> getTrendingRecipes({int limit = 10}) async {
    if (_useMockData) {
      return _getMockTrendingRecipes(limit);
    }

    try {
      final response = await _dio.get(
        '/api/v1/recipes/trending',
        queryParameters: {
          'limit': limit.toString(),
        },
      );

      if (response.statusCode == 200) {
        final data = response.data as Map<String, dynamic>;
        return (data['trending_recipes'] as List<dynamic>)
            .map((recipe) => Recipe.fromJson(recipe as Map<String, dynamic>))
            .toList();
      } else {
        throw Exception('Failed to get trending recipes: ${response.statusCode}');
      }
    } on DioException catch (e) {
      debugPrint('Trending recipes error: ${e.message}');
      return _getMockTrendingRecipes(limit);
    } catch (e) {
      debugPrint('Unexpected error getting trending recipes: $e');
      return _getMockTrendingRecipes(limit);
    }
  }

  /// Get personalized recipe recommendations for the user
  Future<List<Recipe>> getPersonalizedRecommendations({int limit = 10}) async {
    if (_useMockData) {
      return _getMockRecommendations(limit);
    }

    try {
      final response = await _dio.get(
        '/api/v1/recipes/personalized',
        queryParameters: {
          'limit': limit.toString(),
        },
      );

      if (response.statusCode == 200) {
        final data = response.data as Map<String, dynamic>;
        return (data['recommended_recipes'] as List<dynamic>)
            .map((recipe) => Recipe.fromJson(recipe as Map<String, dynamic>))
            .toList();
      } else {
        throw Exception('Failed to get personalized recommendations: ${response.statusCode}');
      }
    } on DioException catch (e) {
      debugPrint('Personalized recommendations error: ${e.message}');
      return _getMockRecommendations(limit);
    } catch (e) {
      debugPrint('Unexpected error getting personalized recommendations: $e');
      return _getMockRecommendations(limit);
    }
  }

  /// Get recipes by meal type (convenience method)
  Future<RecipeSearchResult> getRecipesByMealType(
    String mealType, {
    int page = 1,
    int limit = 20,
  }) async {
    final filters = RecipeFilters(mealType: mealType);
    return searchRecipes(
      filters: filters,
      page: page,
      limit: limit,
    );
  }

  /// Get recipes by cuisine type (convenience method)
  Future<RecipeSearchResult> getRecipesByCuisine(
    String cuisineType, {
    int page = 1,
    int limit = 20,
  }) async {
    final filters = RecipeFilters(cuisineType: cuisineType);
    return searchRecipes(
      filters: filters,
      page: page,
      limit: limit,
    );
  }

  /// Get recipes by dietary restrictions (convenience method)
  Future<RecipeSearchResult> getRecipesByDietaryRestrictions(
    List<String> dietaryRestrictions, {
    int page = 1,
    int limit = 20,
  }) async {
    final filters = RecipeFilters(dietaryRestrictions: dietaryRestrictions);
    return searchRecipes(
      filters: filters,
      page: page,
      limit: limit,
    );
  }

  /// Get quick recipes (under 30 minutes)
  Future<RecipeSearchResult> getQuickRecipes({
    int page = 1,
    int limit = 20,
  }) async {
    final filters = RecipeFilters(maxPrepTime: 30);
    return searchRecipes(
      filters: filters,
      page: page,
      limit: limit,
      sortBy: 'prep_time',
      sortOrder: 'asc',
    );
  }

  /// Get budget-friendly recipes
  Future<RecipeSearchResult> getBudgetFriendlyRecipes({
    int page = 1,
    int limit = 20,
  }) async {
    final filters = RecipeFilters(maxCostUsd: 5.0);
    return searchRecipes(
      filters: filters,
      page: page,
      limit: limit,
      sortBy: 'cost',
      sortOrder: 'asc',
    );
  }

  /// Get beginner-friendly recipes
  Future<RecipeSearchResult> getBeginnerRecipes({
    int page = 1,
    int limit = 20,
  }) async {
    final filters = RecipeFilters(difficultyLevel: 'beginner');
    return searchRecipes(
      filters: filters,
      page: page,
      limit: limit,
    );
  }

  // MOCK DATA METHODS
  List<Recipe> _getMockRecipes() {
    return [
      Recipe(
        id: 'recipe-1',
        name: 'Classic Spaghetti Carbonara',
        description: 'A traditional Italian pasta dish with eggs, cheese, and pancetta',
        ingredients: [
          Ingredient(name: 'Spaghetti', quantity: '400', unit: 'g'),
          Ingredient(name: 'Pancetta', quantity: '150', unit: 'g'),
          Ingredient(name: 'Eggs', quantity: '4', unit: 'large'),
          Ingredient(name: 'Parmesan cheese', quantity: '100', unit: 'g'),
          Ingredient(name: 'Black pepper', quantity: '1', unit: 'tsp'),
        ],
        instructions: 'Cook pasta, prepare sauce, combine and serve',
        cuisineType: 'Italian',
        mealType: 'dinner',
        prepTimeMinutes: 15,
        cookTimeMinutes: 20,
        nutritionalInfo: NutritionalInfo(calories: 520, protein: 28, carbs: 65, fat: 18),
        estimatedCostUsd: 850, // $8.50 in cents
        difficultyLevel: 'medium',
        servings: 4,
        isActive: true,
      ),
      Recipe(
        id: 'recipe-2',
        name: 'Chicken Tikka Masala',
        description: 'Tender chicken in a creamy, spiced tomato sauce',
        ingredients: [
          Ingredient(name: 'Chicken breast', quantity: '500', unit: 'g'),
          Ingredient(name: 'Yogurt', quantity: '200', unit: 'ml'),
          Ingredient(name: 'Tomato sauce', quantity: '400', unit: 'ml'),
          Ingredient(name: 'Heavy cream', quantity: '100', unit: 'ml'),
          Ingredient(name: 'Garam masala', quantity: '2', unit: 'tsp'),
        ],
        instructions: 'Marinate chicken, cook, prepare sauce, combine',
        cuisineType: 'Indian',
        mealType: 'dinner',
        prepTimeMinutes: 30,
        cookTimeMinutes: 25,
        nutritionalInfo: NutritionalInfo(calories: 420, protein: 35, carbs: 12, fat: 25),
        estimatedCostUsd: 1200, // $12.00 in cents
        difficultyLevel: 'medium',
        servings: 4,
        isActive: true,
      ),
      Recipe(
        id: 'recipe-3',
        name: 'Avocado Toast',
        description: 'Simple and healthy breakfast with creamy avocado',
        ingredients: [
          Ingredient(name: 'Bread slices', quantity: '2', unit: 'pieces'),
          Ingredient(name: 'Avocado', quantity: '1', unit: 'large'),
          Ingredient(name: 'Lemon juice', quantity: '1', unit: 'tbsp'),
          Ingredient(name: 'Salt', quantity: 'to taste', unit: ''),
          Ingredient(name: 'Red pepper flakes', quantity: 'pinch', unit: ''),
        ],
        instructions: 'Toast bread, mash avocado, spread and season',
        cuisineType: 'American',
        mealType: 'breakfast',
        prepTimeMinutes: 5,
        cookTimeMinutes: 2,
        nutritionalInfo: NutritionalInfo(calories: 280, protein: 8, carbs: 30, fat: 18),
        estimatedCostUsd: 350, // $3.50 in cents
        difficultyLevel: 'easy',
        servings: 1,
        isActive: true,
      ),
      Recipe(
        id: 'recipe-4',
        name: 'Beef Stir Fry',
        description: 'Quick and flavorful Asian-inspired dish',
        ingredients: [
          Ingredient(name: 'Beef strips', quantity: '300', unit: 'g'),
          Ingredient(name: 'Mixed vegetables', quantity: '400', unit: 'g'),
          Ingredient(name: 'Soy sauce', quantity: '3', unit: 'tbsp'),
          Ingredient(name: 'Garlic', quantity: '3', unit: 'cloves'),
          Ingredient(name: 'Ginger', quantity: '1', unit: 'inch'),
        ],
        instructions: 'Heat oil, cook beef, add vegetables, season and toss',
        cuisineType: 'Asian',
        mealType: 'dinner',
        prepTimeMinutes: 15,
        cookTimeMinutes: 10,
        nutritionalInfo: NutritionalInfo(calories: 350, protein: 30, carbs: 15, fat: 20),
        estimatedCostUsd: 950, // $9.50 in cents
        difficultyLevel: 'easy',
        servings: 3,
        isActive: true,
      ),
      Recipe(
        id: 'recipe-5',
        name: 'Caesar Salad',
        description: 'Crisp romaine lettuce with classic Caesar dressing',
        ingredients: [
          Ingredient(name: 'Romaine lettuce', quantity: '1', unit: 'head'),
          Ingredient(name: 'Parmesan cheese', quantity: '50', unit: 'g'),
          Ingredient(name: 'Croutons', quantity: '100', unit: 'g'),
          Ingredient(name: 'Caesar dressing', quantity: '4', unit: 'tbsp'),
        ],
        instructions: 'Wash lettuce, add dressing, top with cheese and croutons',
        cuisineType: 'American',
        mealType: 'lunch',
        prepTimeMinutes: 10,
        cookTimeMinutes: 0,
        nutritionalInfo: NutritionalInfo(calories: 220, protein: 12, carbs: 18, fat: 12),
        estimatedCostUsd: 600, // $6.00 in cents
        difficultyLevel: 'easy',
        servings: 2,
        isActive: true,
      ),
    ];
  }

  RecipeSearchResult _getMockSearchResult(String searchQuery, RecipeFilters? filters, int page, int limit, String sortBy, String sortOrder) {
    final allRecipes = _getMockRecipes();
    
    // Simple filtering
    var filteredRecipes = allRecipes.where((recipe) {
      if (searchQuery.isNotEmpty) {
        return recipe.name.toLowerCase().contains(searchQuery.toLowerCase()) ||
               recipe.description?.toLowerCase().contains(searchQuery.toLowerCase()) == true;
      }
      return true;
    }).toList();

    // Apply filters
    if (filters != null) {
      if (filters.mealType != null) {
        filteredRecipes = filteredRecipes.where((r) => r.mealType == filters.mealType).toList();
      }
      if (filters.cuisineType != null) {
        filteredRecipes = filteredRecipes.where((r) => r.cuisineType == filters.cuisineType).toList();
      }
      if (filters.difficultyLevel != null) {
        filteredRecipes = filteredRecipes.where((r) => r.difficultyLevel == filters.difficultyLevel).toList();
      }
    }

    // Simple pagination
    final startIndex = (page - 1) * limit;
    final endIndex = startIndex + limit;
    final paginatedRecipes = filteredRecipes.skip(startIndex).take(limit).toList();

    return RecipeSearchResult(
      recipes: paginatedRecipes,
      pagination: RecipePagination(
        page: page,
        limit: limit,
        totalCount: filteredRecipes.length,
        totalPages: (filteredRecipes.length / limit).ceil(),
        hasNext: endIndex < filteredRecipes.length,
        hasPrevious: page > 1,
      ),
      filtersApplied: filters?.toQueryParams() ?? {},
      searchQuery: searchQuery,
      sort: RecipeSort(
        sortBy: sortBy,
        sortOrder: sortOrder,
      ),
    );
  }

  Recipe _getMockRecipe(String recipeId) {
    final recipes = _getMockRecipes();
    return recipes.firstWhere(
      (recipe) => recipe.id == recipeId,
      orElse: () => recipes.first,
    );
  }

  FilterOptions _getMockFilterOptions() {
    return FilterOptions(
      mealTypes: ['breakfast', 'lunch', 'dinner', 'snack'],
      cuisineTypes: ['Italian', 'Indian', 'American', 'Asian', 'Mexican'],
      difficultyLevels: ['easy', 'medium', 'hard'],
      dietaryRestrictions: ['vegetarian', 'vegan', 'gluten-free', 'dairy-free'],
      timeRanges: [
        TimeRange(label: 'Quick (< 15 min)', maxMinutes: 15),
        TimeRange(label: 'Medium (15-30 min)', minMinutes: 15, maxMinutes: 30),
        TimeRange(label: 'Long (> 30 min)', minMinutes: 30),
      ],
      costRanges: [
        CostRange(label: 'Budget (< \$5)', maxUsd: 5.0),
        CostRange(label: 'Moderate (\$5-\$10)', minUsd: 5.0, maxUsd: 10.0),
        CostRange(label: 'Premium (> \$10)', minUsd: 10.0),
      ],
    );
  }

  List<String> _getMockSearchSuggestions(String query, int limit) {
    final suggestions = [
      'Spaghetti Carbonara',
      'Chicken Tikka Masala',
      'Avocado Toast',
      'Beef Stir Fry',
      'Caesar Salad',
      'Chicken Curry',
      'Pasta Primavera',
      'Fish Tacos',
      'Greek Salad',
      'Mushroom Risotto',
    ];
    
    return suggestions
        .where((s) => s.toLowerCase().contains(query.toLowerCase()))
        .take(limit)
        .toList();
  }

  List<Recipe> _getMockTrendingRecipes(int limit) {
    return _getMockRecipes().take(limit).toList();
  }

  List<Recipe> _getMockRecommendations(int limit) {
    return _getMockRecipes().reversed.take(limit).toList();
  }
} 

File: functions/src/services/recipe_discovery_service.py
"""
Recipe Discovery Service
Business logic for recipe search, filtering, and recommendations
"""

import logging
from typing import List, Dict, Any, Optional, NamedTuple
from dataclasses import dataclass
from sqlalchemy import and_, or_, desc, asc
from sqlalchemy.orm import Session
import math

from core.models.recipe import Recipe
from core.models.user import User
from data_access.recipe_repository import RecipeRepository
from data_access.user_repository import UserRepository
from core.exceptions import ValidationError
from data_access.database import db

logger = logging.getLogger(__name__)

@dataclass
class SearchResult:
    """Container for search results with pagination info"""
    recipes: List[Recipe]
    page: int
    limit: int
    total_count: int
    total_pages: int
    has_next: bool
    has_previous: bool
    filters_applied: Dict[str, Any]

class RecipeDiscoveryService:
    """Service for recipe discovery, search, and recommendations"""
    
    def __init__(self, session: Optional[Session] = None):
        """Initialize service with optional database session"""
        self.session = session or db.session
        self.recipe_repository = RecipeRepository(self.session)
        self.user_repository = UserRepository()
    
    def search_recipes(self, search_query: str = "", filters: Optional[Dict[str, Any]] = None,
                      page: int = 1, limit: int = 20, sort_by: str = "name", 
                      sort_order: str = "asc", user_id: Optional[str] = None) -> SearchResult:
        """
        Search recipes with comprehensive filtering and pagination
        
        Args:
            search_query: Text to search for in recipe names, descriptions, ingredients
            filters: Dictionary of filter criteria
            page: Page number (1-based)
            limit: Number of results per page
            sort_by: Field to sort by (name, prep_time, cost, difficulty, created_at)
            sort_order: Sort direction (asc, desc)
            user_id: User ID for personalization
            
        Returns:
            SearchResult with recipes and pagination info
        """
        try:
            filters = filters or {}
            
            # Start with base query for active recipes
            query = self.session.query(Recipe).filter(Recipe.is_active == True)
            
            # Apply text search if provided
            if search_query:
                search_terms = search_query.lower().split()
                for term in search_terms:
                    # Search in name, description, and ingredients
                    search_filter = or_(
                        Recipe.name.ilike(f"%{term}%"),
                        Recipe.description.ilike(f"%{term}%"),
                        Recipe.ingredients.cast(db.String).ilike(f"%{term}%")
                    )
                    query = query.filter(search_filter)
            
            # Apply filters
            query = self._apply_filters(query, filters)
            
            # Get total count before pagination
            total_count = query.count()
            
            # Apply sorting
            query = self._apply_sorting(query, sort_by, sort_order)
            
            # Apply pagination
            offset = (page - 1) * limit
            recipes = query.offset(offset).limit(limit).all()
            
            # Calculate pagination info
            total_pages = math.ceil(total_count / limit) if total_count > 0 else 1
            has_next = page < total_pages
            has_previous = page > 1
            
            # Apply personalization scoring if user provided
            if user_id and recipes:
                recipes = self._apply_personalization(recipes, user_id)
            
            logger.info(f"Recipe search completed: {len(recipes)} recipes found (page {page}/{total_pages})")
            
            return SearchResult(
                recipes=recipes,
                page=page,
                limit=limit,
                total_count=total_count,
                total_pages=total_pages,
                has_next=has_next,
                has_previous=has_previous,
                filters_applied=filters
            )
            
        except Exception as e:
            logger.error(f"Error in recipe search: {str(e)}")
            raise ValidationError(f"Recipe search failed: {str(e)}")
    
    def _apply_filters(self, query, filters: Dict[str, Any]):
        """Apply various filters to the recipe query"""
        
        # Meal type filter
        if 'meal_type' in filters:
            query = query.filter(Recipe.meal_type == filters['meal_type'])
        
        # Cuisine type filter
        if 'cuisine_type' in filters:
            query = query.filter(Recipe.cuisine_type.ilike(f"%{filters['cuisine_type']}%"))
        
        # Difficulty level filter
        if 'difficulty_level' in filters:
            query = query.filter(Recipe.difficulty_level == filters['difficulty_level'])
        
        # Preparation time filter
        if 'max_prep_time' in filters:
            query = query.filter(Recipe.prep_time_minutes <= filters['max_prep_time'])
        
        # Cost filters
        if 'min_cost_usd' in filters:
            min_cost_cents = int(filters['min_cost_usd'] * 100)
            query = query.filter(Recipe.estimated_cost_usd >= min_cost_cents)
        
        if 'max_cost_usd' in filters:
            max_cost_cents = int(filters['max_cost_usd'] * 100)
            query = query.filter(Recipe.estimated_cost_usd <= max_cost_cents)
        
        # Dietary restrictions - this needs to be applied after the query
        # since it requires Python logic in the Recipe model
        return query
    
    def _apply_sorting(self, query, sort_by: str, sort_order: str):
        """Apply sorting to the recipe query"""
        
        # Map sort fields to actual model attributes
        sort_mapping = {
            'name': Recipe.name,
            'prep_time': Recipe.prep_time_minutes,
            'cost': Recipe.estimated_cost_usd,
            'difficulty': Recipe.difficulty_level,
            'created_at': Recipe.created_at
        }
        
        if sort_by not in sort_mapping:
            sort_by = 'name'  # Default fallback
        
        sort_field = sort_mapping[sort_by]
        
        if sort_order == 'desc':
            query = query.order_by(desc(sort_field))
        else:
            query = query.order_by(asc(sort_field))
        
        return query
    
    def _apply_personalization(self, recipes: List[Recipe], user_id: str) -> List[Recipe]:
        """Apply personalization scoring to recipes based on user preferences"""
        try:
            # Get user and their preferences
            user = self.user_repository.get_user_by_id(user_id)
            if not user:
                return recipes
            
            # For now, just return recipes as-is
            # In future iterations, this could:
            # 1. Score recipes based on user's past preferences
            # 2. Consider user's dietary restrictions
            # 3. Factor in user's cooking skill level
            # 4. Use ML models for recommendation scoring
            
            return recipes
            
        except Exception as e:
            logger.warning(f"Personalization failed for user {user_id}: {str(e)}")
            return recipes
    
    def get_recipe_details(self, recipe_id: str, user_id: Optional[str] = None) -> Optional[Recipe]:
        """Get detailed information for a specific recipe"""
        try:
            recipe = self.recipe_repository.get_recipe_by_id(recipe_id)
            
            # Log recipe view for analytics (optional)
            if recipe and user_id:
                logger.info(f"User {user_id} viewed recipe {recipe_id}")
            
            return recipe
            
        except Exception as e:
            logger.error(f"Error getting recipe details {recipe_id}: {str(e)}")
            raise ValidationError(f"Failed to get recipe details: {str(e)}")
    
    def get_filter_options(self) -> Dict[str, List[str]]:
        """Get available filter options based on existing recipes"""
        try:
            # Query distinct values for filter options
            meal_types = self.session.query(Recipe.meal_type).distinct().filter(
                and_(Recipe.meal_type.isnot(None), Recipe.is_active == True)
            ).all()
            
            cuisine_types = self.session.query(Recipe.cuisine_type).distinct().filter(
                and_(Recipe.cuisine_type.isnot(None), Recipe.is_active == True)
            ).all()
            
            difficulty_levels = self.session.query(Recipe.difficulty_level).distinct().filter(
                and_(Recipe.difficulty_level.isnot(None), Recipe.is_active == True)
            ).all()
            
            # Convert tuples to lists and filter out None values
            meal_type_list = sorted([mt[0] for mt in meal_types if mt[0]])
            cuisine_type_list = sorted([ct[0] for ct in cuisine_types if ct[0]])
            difficulty_level_list = sorted([dl[0] for dl in difficulty_levels if dl[0]])
            
            return {
                'meal_types': meal_type_list,
                'cuisine_types': cuisine_type_list,
                'difficulty_levels': difficulty_level_list,
                'dietary_restrictions': [
                    'vegan', 'vegetarian', 'gluten-free', 'dairy-free', 'nut-free', 'keto'
                ],
                'time_ranges': [
                    {'label': '≤15 min', 'max_minutes': 15},
                    {'label': '15-30 min', 'min_minutes': 15, 'max_minutes': 30},
                    {'label': '30-60 min', 'min_minutes': 30, 'max_minutes': 60},
                    {'label': '>60 min', 'min_minutes': 60}
                ],
                'cost_ranges': [
                    {'label': 'Budget-friendly', 'max_usd': 5.00},
                    {'label': 'Moderate', 'min_usd': 5.00, 'max_usd': 15.00},
                    {'label': 'Premium', 'min_usd': 15.00}
                ]
            }
            
        except Exception as e:
            logger.error(f"Error getting filter options: {str(e)}")
            raise ValidationError(f"Failed to get filter options: {str(e)}")
    
    def get_search_suggestions(self, query: str, limit: int = 10, 
                             user_id: Optional[str] = None) -> List[str]:
        """Get search suggestions based on partial query"""
        try:
            if len(query) < 2:
                return []
            
            # Search for recipe names that start with or contain the query
            recipes = self.session.query(Recipe.name).filter(
                and_(
                    Recipe.name.ilike(f"%{query}%"),
                    Recipe.is_active == True
                )
            ).limit(limit).all()
            
            suggestions = [recipe[0] for recipe in recipes]
            
            # Add some common ingredient suggestions
            common_ingredients = [
                'chicken', 'beef', 'salmon', 'pasta', 'rice', 'vegetables',
                'garlic', 'onion', 'tomato', 'cheese', 'herbs', 'spices'
            ]
            
            ingredient_suggestions = [
                ing for ing in common_ingredients 
                if query.lower() in ing.lower() and ing not in suggestions
            ]
            
            # Combine and limit results
            all_suggestions = suggestions + ingredient_suggestions
            return all_suggestions[:limit]
            
        except Exception as e:
            logger.error(f"Error getting search suggestions: {str(e)}")
            return []
    
    def get_trending_recipes(self, limit: int = 10, user_id: Optional[str] = None) -> List[Recipe]:
        """Get trending/popular recipes"""
        try:
            # For now, return recently created recipes
            # In a real system, this would be based on view counts, ratings, etc.
            recipes = self.session.query(Recipe).filter(
                Recipe.is_active == True
            ).order_by(desc(Recipe.created_at)).limit(limit).all()
            
            return recipes
            
        except Exception as e:
            logger.error(f"Error getting trending recipes: {str(e)}")
            return []
    
    def get_personalized_recommendations(self, user_id: str, limit: int = 10) -> List[Recipe]:
        """Get personalized recipe recommendations for a user"""
        try:
            user = self.user_repository.get_user_by_id(user_id)
            if not user:
                # Fallback to trending recipes
                return self.get_trending_recipes(limit)
            
            # Get user's dietary restrictions
            dietary_restrictions = getattr(user, 'dietary_restrictions', []) or []
            
            # Start with recipes that match dietary restrictions
            query = self.session.query(Recipe).filter(Recipe.is_active == True)
            
            # For now, just filter by dietary restrictions and return recent recipes
            # In future iterations, this could use ML models, user behavior, etc.
            recipes = query.order_by(desc(Recipe.created_at)).limit(limit * 2).all()
            
            # Filter by dietary restrictions in Python
            if dietary_restrictions:
                filtered_recipes = [
                    recipe for recipe in recipes 
                    if recipe.matches_dietary_restrictions(dietary_restrictions)
                ]
                return filtered_recipes[:limit]
            
            return recipes[:limit]
            
        except Exception as e:
            logger.error(f"Error getting personalized recommendations for user {user_id}: {str(e)}")
            # Fallback to trending recipes
            return self.get_trending_recipes(limit) 